jumanji/environments/packing/bin_pack/env.py

```diff
@@ -739,6 +743,148 @@ def _get_intersections_dict(
+                    ~res.is_empty()
+                )
+                return carry, res
+
+            def delete_merged_ems_and_add_new_ems(
+                carry: EMS, merged_ems_ind: Tuple[chex.Array]
+            ) -> Tuple[EMS, Any]:
+                carry = jax.lax.cond(
+                    is_merged_ems[merged_ems_ind],
+                    lambda ems_arr, ems_mask: (
+                        tree_add_element(
+                            tree_add_element(
+                                ems_arr,
+                                merged_ems_indices[0][merged_ems_ind],
+                                tree_slice(
+                                    merged_ems,
+                                    merged_ems_indices[0][merged_ems_ind] * mask_shape
+                                    + merged_ems_indices[1][merged_ems_ind],
+                                ),
+                            ),
+                            merged_ems_indices[1][merged_ems_ind],
+                            empty_ems,
+                        ),
+                        ems_mask.at[merged_ems_indices[0][merged_ems_ind]]
+                        .set(True)
+                        .at[merged_ems_indices[1][merged_ems_ind]]
+                        .set(False),
+                    ),
+                    lambda *_: _,
+                    *(carry),
+                )
+                return carry, None
+
+            mask, same_x = compute_merge_mask((ems_arr, ems_mask))
+
+            mask_shape = mask.shape[0]
+            init_val = jnp.full_like(mask, jnp.nan)
+
+            same_x = same_x.flatten()
+            mask = mask.flatten()
+
+            # Construct the new ems from merging previous ones
+            # is_merged_ems[i,j] contains a boolean specifying if the two ems at i and j
+            # have been merged
+            is_merged_ems, merged_ems = jax.lax.scan(
+                scan_and_merge, init_val, jnp.arange(len(mask))
+            )
+            is_merged_ems = is_merged_ems.flatten()
+            keys = is_merged_ems.argsort()[::-1]
+            is_merged_ems = is_merged_ems.sort()[::-1]
+            _, merged_ems_indices = jax.lax.scan(
+                lambda carry, key: (carry, (key // mask_shape, key % mask_shape)),
+                jnp.arange(len(ems_arr.x1)),
+                keys,
+            )
+            is_merged_ems = is_merged_ems[: len(ems_arr.x1)]
+            merged_ems_indices = (
+                merged_ems_indices[0][: len(ems_arr.x1)],
+                merged_ems_indices[1][: len(ems_arr.x1)],
+            )
+            (new_ems, new_ems_mask), _ = jax.lax.scan(
+                delete_merged_ems_and_add_new_ems,
+                (ems_arr, ems_mask),
+                jnp.arange(len(ems_arr.x1)),
+            )
+
+            return new_ems, new_ems_mask
+
+        state.ems, state.ems_mask = jax.lax.while_loop(
+            lambda ems: jnp.any(compute_merge_mask(ems)[0]),
```

Comment:
Maybe replace "ems" in the lambda function with something else to make sure it is clear that it is a tuple and not an ems type that's being passed to the compute_merge_mask function.

Reply:
Done ✅